[JIT] Attempt to handle static C++ objects construction/destruction

tlepley-cadence opened the pull request:
Summary: (very) experimental code to solve the handling of static C++ objects in libjit.

Fixes #3839 

@opti-mix

tlepley-cadence commented:
The compilation fails in the CI. Which LLVM version do you use in the CI ? 
I'm currently using LLVM 7 locally.

opti-mix commented:
@tlepley-cadence 

> The compilation fails in the CI. Which LLVM version do you use in the CI ? I'm currently using LLVM 7 locally.

We use LLVM 8.0.1 on CI.

tlepley-cadence commented:
> My main concern is: How much do these LLVM APIs depend on the version of ORC JIT APIs? Are they stable? Or do they change as often as the ORC JIT APIs? 

As far as I understood, the entire GlowJIT code is currently based on the ORCv1 API and the extra 'utility' API I'm using in the MR is consistent since ORCv1 too. I think that the ORCv1 API is stable. For LLVM 8, just few classes now have the 'Legacy' prefix (you guys had already to do this change for other classes in GlowJIT when adding the support for LLVM 8).

Note that ORCv1 is in the process of being deprecated. So I guess that Glow will need at some point to migrate to the ORCv2 API. But this is of course not in the scope of the MR :-)

tlepley-cadence commented on lib/LLVMIRCodeGen/GlowJIT.cpp:
```diff
@@ -130,16 +130,35 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
               return JITSymbol(SymAddr, JITSymbolFlags::Exported);
+            // The symbol was not resolved. Dump an error message and return a
+            // 'symbol not found' JITSymbol object (nullptr).
+            llvm::errs() << "JIT: Error resolving symbol '" << Name << "'\n";
```
The problem is that when such case happens, it doesn't fail immediately. The error will happen later when trying to get the address of the 'jitmain' symbol (in LLVMCompiledFunction.cpp). We then get the "Error getting address" that doesn't tell us anything about what is the actual problem.

tlepley-cadence replied:
I enclosed this into  `#ifndef NDEBUG`. I think it's important to keep it for the reason explained before. An we don't want to fail immediately because multiple symbols may not be resolved, so we want to capture all in one shot.

opti-mix replied:
@tlepley-cadence I'd suggest you just put this inside `DEBUG_GLOW`. This way you can always enable it via a command-line option.

opti-mix replied:
@tlepley-cadence Ping on this one. Please use `DEBUG_GLOW` instead of ifdefs

tlepley-cadence replied:
Done.

tlepley-cadence commented on lib/LLVMIRCodeGen/GlowJIT.cpp:
```diff
@@ -160,21 +179,58 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
+GlowJIT::~GlowJIT() {
+  // Run any destructors registered with __cxa_atexit.
+  cxxoverride_.runDestructors(); // Run any IR destructors.
+  // Run any IR destructors.
```
Fixed

tlepley-cadence commented on lib/LLVMIRCodeGen/GlowJIT.cpp:
```diff
@@ -160,21 +179,58 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
+                                                          K);
+  cantFail(CtorRunner.runViaLayer(compileLayer_));
+
+  IRStaticDestructorRunners.emplace_back(std::move(DtorNames), K);
```
Done

tlepley-cadence commented on lib/LLVMIRCodeGen/GlowJIT.cpp:
```diff
@@ -160,21 +179,58 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
+
+  orc::CtorDtorRunner<decltype(compileLayer_)> CtorRunner(std::move(CtorNames),
+                                                          K);
+  cantFail(CtorRunner.runViaLayer(compileLayer_));
```
Done.

tlepley-cadence commented on lib/LLVMIRCodeGen/GlowJIT.cpp:
```diff
@@ -160,21 +179,58 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
+  // we hand over ownership of the module to the JIT.
+  // See:
+  // https://github.com/microsoft/DirectXShaderCompiler/blob/master/tools/lli/OrcLazyJIT.h
+  std::vector<std::string> CtorNames, DtorNames;
```
Done.

tlepley-cadence commented on lib/LLVMIRCodeGen/GlowJIT.cpp:
```diff
@@ -130,16 +130,35 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
+            // Some symbols are overriden, in particular __dso_handle and
+            // __cxa_atexit
+            if (auto overridenSym = cxxoverride_.searchOverrides(Name)) {
+              llvm::errs() << "JIT: Found overriden symbol '" << Name << "'\n";
```
Removed. This message is not needed.

tlepley-cadence commented on include/glow/LLVMIRCodeGen/GlowJIT.h:
```diff
@@ -61,9 +63,13 @@ class GlowJIT {
 #endif
+  std::vector<llvm::orc::CtorDtorRunner<decltype(compileLayer_)>>
+      IRStaticDestructorRunners;
+  std::string mangle(const std::string &Name);
```
Done

tlepley-cadence commented on include/glow/LLVMIRCodeGen/GlowJIT.h:
```diff
@@ -52,6 +52,8 @@ class GlowJIT {
 #else
   std::shared_ptr<SymbolStringPool> SSP_;
   ExecutionSession ES_;
+  llvm::orc::LocalCXXRuntimeOverrides cxxoverride_;
```
Done

tlepley-cadence commented on include/glow/LLVMIRCodeGen/GlowJIT.h:
```diff
@@ -61,9 +63,13 @@ class GlowJIT {
   LegacyRTDyldObjectLinkingLayer objectLayer_;
   LegacyIRCompileLayer<decltype(objectLayer_), SimpleCompiler> compileLayer_;
 #endif
+  std::vector<llvm::orc::CtorDtorRunner<decltype(compileLayer_)>>
```
Done

tlepley-cadence commented on include/glow/LLVMIRCodeGen/GlowJIT.h:
```diff
@@ -61,9 +63,13 @@ class GlowJIT {
   LegacyIRCompileLayer<decltype(objectLayer_), SimpleCompiler> compileLayer_;
 #endif
+  std::vector<llvm::orc::CtorDtorRunner<decltype(compileLayer_)>>
+      IRStaticDestructorRunners;
```
Done

opti-mix commented:
> As far as I understood, the entire GlowJIT code is currently based on the ORCv1 API and the extra 'utility' API I'm using in this PR is consistent since using ORCv1 too. I think that the ORCv1 API is stable. For LLVM 8, just few classes now have the 'Legacy' prefix (you guys had already to do this change for other classes in GlowJIT when adding the support for LLVM 8).
Note that ORCv1 is in the process of being deprecated. So I guess that Glow will need at some point to migrate to the ORCv2 API. But this is of course not in the scope of the MR :-)

Yep. We are planning to migrate to ORCv2 APIs in the near future. Thus, having this PR support both ORCv1 and ORCv2 would make our life easier. 

But you are right that it could be done in a separate PR. Though we would really appreciate if it could be a part of this PR 🤣

tlepley-cadence commented:
> Ideally, we'd like to support this functionality on LLVM7, LLVM8 and LLVM9, but I'm afraid it will require a lot of #ifdefs ;-)

This is done. I've tested with LLVM 7, 8 and 9. The ifdef actually only concerns LLVM  7 (they added the 'Legacy' prefix in LLVM 8 and kept the same API in 9). I could take advantage of some already existing #ifdef :)

tlepley-cadence commented:
Which version of clangformat do you use for the formating ? I've used LLVM/CLANG 8 and it fails the CI checking. It seems that you still use Clang 7. Do you confirm ?

opti-mix commented:
> It seems that you still use Clang 7. Do you confirm ?

Yes. We currently use clang-format from Clang 7.

opti-mix commented:
@tlepley-cadence Looking pretty good now!

> And BTW could you try to add a small test for this feature? Something that uses the CPU backend and plays with C++ constructors/destructors?

Any chance you'd be able to add a small test for this functionality?

tlepley-cadence commented:
> And BTW could you try to add a small test for this feature? Something that uses the CPU backend and plays with C++ constructors/destructors?

The simplest way to test this feature would be to add some test code to the CPU libjit (using a global C++ object) but IMO, this is not very clean to patch the CPU libjit. 

Did you have another idea for the testing ? I've not seen any unit test that tests the JIT mechanics.

tlepley-cadence commented:
It seems that there is an CI environment related problem with Pytorch because other PRs are failing the same way.

tlepley-cadence commented:
> Yep. We are planning to migrate to ORCv2 APIs in the near future. Thus, having this PR support both ORCv1 and ORCv2 would make our life easier.
> But you are right that it could be done in a separate PR. Though we would really appreciate if it could be a part of this PR rofl

I think that the migration to ORCv2 needs to be one alltogether. I can't just use ORCv2 for solving the static C++ object problem while the rest of the JIT still uses ORCv1. And I have no plan to migrate the whole GlowJIT to ORCv2, at least at short term and in this PR :-)

tlepley-cadence commented:
@opti-mix locally, the 'small test' I did is to create a 'jittest' external backend in order no to put test code into the CPU or interpreter backend. Do you want me to push this ? or do you have another solution of the jit testing ?

opti-mix commented:
@tlepley-cadence 

> locally, the 'small test' I did is to create a 'jittest' external backend dedicated to testing. In such a way, we don't pollute CPU or interpreter backends with test code. Do you want me to push this ? or do you have another solution for the jit testing ?

Can you describe in a bit more detail what your jittest actually does. Like what C++ constructors/destructors are defined and need to be called when creating/destroying a new module to GlowJIT? I'm just trying to understand how the test looks like, before I can advice on how to proceed. Or may be if you'd push your jittest it will make it more clear.

tlepley-cadence commented:
> Can you describe in a bit more detail what your jittest actually does.

It's very simple:
```
#include <map>
extern std::map<float, float> GlobalMap;
std::map<float, float> GlobalMap;

extern "C" {
/// Simply write a global map to validate that static C++ constructors are
/// correctly called.
void jittest_WriteGlobalMap(float key, float value) {
  GlobalMap[key] = value;
}
```

opti-mix commented:
> Creating a new backend just for testing this looks a bit overkill :-)

Yeah, we have used mock backends defined in tests already. Just search for `MockBackend` in the codebase. 

Could we define a simple mock backend derived from `LLVMBackend` or even `CPUBackend` and then make sure to emit there an LLVM call that would call the `jittest_WriteGlobalMap` you defined above?

tlepley-cadence commented:
> Could we define a simple mock backend derived from `LLVMBackend` or even `CPUBackend` and then make sure to emit there an LLVM call that would call the `jittest_WriteGlobalMap` you defined above?

Actually, my current 'JIT test' external backend is derived from `CPUBackend`. It falls back to `CPUBackend` whenever it gets an instruction it doesn't know. I'm going to look at `MockBackend` to see if it can replace my 'JIT test' backend.

tlepley-cadence commented:
> Could we define a simple mock backend derived from `LLVMBackend` or even `CPUBackend` <...>?
I did some experiments creating a mock backend and I founf this very limited for JIT related testing. For instance, we can't use ExecutionEngine, can't define test specific nodes/instructions, can't easily create a LLVM bitcode for the libjet. All this requires an actual registered backend.

What I did for now is to upstream my `JITTest` 'external backend' that is registered as an actual backend and that has its own nodes and tests. It is derived from the CPU backend. Like I told earlier, it's a bit overkill, but it may be reused extended later for other JIT related testing.

opti-mix commented:
@tlepley-cadence Thanks for providing the code for the test.

I managed to build your PR on MacOS, but I get the following errors when I run the test:
```
JIT: Error resolving symbol '__ZNSt3__16__treeINS_12__value_typeIffEENS_19__map_value_compareIfS2_NS_4lessIfEELb1EEENS_9allocatorIS2_EEE7destroyEPNS_11__tree_nodeIS2_PvEE'
JIT: Error resolving symbol '__ZNSt3__13mapIffNS_4lessIfEENS_9allocatorINS_4pairIKffEEEEED1Ev'
JIT: Error resolving symbol '__ZNSt3__16__treeINS_12__value_typeIffEENS_19__map_value_compareIfS2_NS_4lessIfEELb1EEENS_9allocatorIS2_EEE7destroyEPNS_11__tree_nodeIS2_PvEE'
JIT: Error resolving symbol '__ZNSt3__13mapIffNS_4lessIfEENS_9allocatorINS_4pairIKffEEEEED1Ev'
```

Do you see the same errors?

opti-mix commented:
@tlepley-cadence The interesting thing about those MacOS errors I reported above is that it seems like the symbol is in the module if I dump it, but it cannot be found.

opti-mix commented:
Ping!

tlepley-cadence commented:
@opti-mix sorry for the delay, I was travelling.

The fix I did solved the handling of static constructor/destructors (__dso_handle).

But it seems there are additional symbol resolution issues (at least with MacOS and Windows). For instance, when simply using C++ strings and stream in libjit (without using any global C++ object), it works well under Linux, but we also got some symbol resolution problems under Windows.

Do you mean that in your case, the symbols that fail resolution are defined in the libjit module itself ? Are you sure they are not simply referenced, rather than being defined ?

opti-mix commented:
@tlepley-cadence 

> Do you mean that in your case, the symbols that fail resolution are defined in the libjit module itself ? Are you sure they are not simply referenced, rather than being defined ?

OK. I looked a bit closer. I seem to observe the following:
- The two mentioned symbols are present in the libjit mode before it is optimized. They are originally reachable from `__cxx_global_var_init`, which is reachable from the global initializer generated for  the source file, which in turn is listed in `llvm.global_ctors`.

- After the optimization is done, the global initializer generated for  the source file is still listed in `llvm.global_ctors`. But it does not invoke `__cxx_global_var_init` anymore. Instead is seems like it inlined it and multiple levels of functions called from it. So, some of these functions are not referenced anymore. But interestingly enough, some of them are not removed by the optimizer. Probably, because they have a wrong linkage or something like this. As a result, they are still in the LLVMModule, but are not used and no address is generated for them. These symbols are still recognized as global constructors or destructors due to their names. When this module is being added to the GlowJIT, the JIT cannot resolve them properly, as they don't have any addresses assigned to them somehow (probably because they are not referenced).

opti-mix commented:
@tlepley-cadence OK some further investigation has shown that ORCJit doesn't like `linkonce_odr` linkage. But the following small snippet in Pipeline.cpp, in optimizeLLVMModule seems to help a lot:
```c++
    FF.removeFnAttr(llvm::Attribute::AttrKind::NoInline);
    if (!FF.isDeclaration() && FF.isLinkOnceLinkage(FF.getLinkage())) {
      FF.setLinkage(llvm::GlobalValue::LinkageTypes::InternalLinkage);
    }

```

opti-mix commented:
@tlepley-cadence Now that the mystery about missing symbols is resolved, I'd like to mention comment about something else.

As you said yourself, your current test case introduces a lot of files and is an overkill. How about something almost equivalent, but much smaller that could live in the same directory with other tests. Here is a sketch:

File CPUJitTestl.cpp:
```c++
/**
 * Copyright (c) Glow Contributors. See CONTRIBUTORS file.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "glow/Backend/Backend.h"
#include "glow/ExecutionEngine/ExecutionEngine.h"
#include "glow/Graph/Graph.h"
#include "glow/Graph/Hook.h"
#include "glow/Graph/Node.h"
#include "glow/Graph/Nodes.h"
#include "glow/Graph/Utils.h"

#include "../../lib/Backends/CPU/CPUBackend.h"
#include "../../lib/Backends/CPU/CPULLVMIRGen.h"

#include "gtest/gtest.h"

using namespace glow;

/// We compile the standard library (libjit) to LLVM bitcode, and then convert
/// that binary data to an include file using an external utility (include-bin).
/// The resulting file is included here to compile the bitcode image into our
/// library.
static const unsigned char libjit_bc[] = {
#include "glow/CPU/mini_libjit_bc.inc"
};
static const size_t libjit_bc_size = sizeof(libjit_bc);

class MockLLVMIRGen : public CPULLVMIRGen {
  virtual void generateLLVMIRForInstr(llvm::IRBuilder<> &builder,
                                      const glow::Instruction *I) override {
    auto key = 1.2f;
    auto *keyValue = emitConstF32(builder, key);
    // Emit library function call.
    auto *F = getFunction("WriteGlobalMap");
    createCall(builder, F, {keyValue, keyValue});
  }
  bool
  canBePartOfDataParallelKernel(const glow::Instruction *I) const override {
    return false;
  }
  // void generateLLVMIRForModule(llvm::IRBuilder<> &builder) override {
  //   // // Add a declaration of libjit_WriteGlobalMap.
  //   // auto *funcTy = llvm::FunctionType::get(builder.getVoidTy(),
  //   // {builder.getFloatTy(), builder.getFloatTy()}, false);
  //   // llvm::Function::Create(funcTy, llvm::Function::ExternalLinkage,
  //   //                        "libjit_WriteGlobalMap", getModule());
  //   CPULLVMIRGen::generateLLVMIRForModule(builder);
  // }

public:
  MockLLVMIRGen(const IRFunction *F, AllocationsInfo &allocationsInfo,
                std::string mainEntryName, llvm::StringRef libjitBC)
      : CPULLVMIRGen(F, allocationsInfo, mainEntryName, libjitBC) {}
};

class MockCPUBackend : public CPUBackend {
  virtual std::unique_ptr<LLVMIRGen>
  createIRGen(const IRFunction *IR,
              AllocationsInfo &allocationsInfo) const override {
    MockLLVMIRGen *irgen =
        new MockLLVMIRGen(IR, allocationsInfo, "", getLibjitBitcode());
    return std::unique_ptr<MockLLVMIRGen>(irgen);
  }
  virtual std::string getBackendName() const override {
    return "MockCPUBackend";
  }

public:
  virtual llvm::StringRef getLibjitBitcode() const override {
    return llvm::StringRef(reinterpret_cast<const char *>(libjit_bc),
                           libjit_bc_size);
  }
  static std::string getName() { return "MockCPUBackend"; }
};

REGISTER_GLOW_BACKEND_FACTORY(MockCPUFactory, MockCPUBackend);

/// Test BundleSaver. This unit test is for code coverage.
TEST(CPUJITTest, testCppConstructors) {
  glow::ExecutionEngine EE("MockCPUBackend");
  Module &M = EE.getModule();
  Function *F = M.createFunction("F");

  // Create a simple graph.
  auto *inputPH = M.createPlaceholder(ElemKind::FloatTy, {1}, "input", false);
  Node *addConst = M.createConstant(ElemKind::FloatTy, {1}, "const");
  Node *addNode = F->createAdd("add", inputPH, addConst);
  auto *saveNode = F->createSave("output", addNode);

  PlaceholderBindings bindings;
  auto *inputT = bindings.allocate(inputPH);
  inputT->getHandle().clear(1.0);
  Tensor expectedT(inputT->getType());
  expectedT.getHandle().clear(1.0);
  auto *outputT = bindings.allocate(saveNode->getPlaceholder());
  EE.compile(CompilationMode::Infer);
  EE.run(bindings);
  EXPECT_TRUE(outputT->isEqual(expectedT));
}
```

and file mini_libjit.cpp:
```c++
/**
 * Copyright (c) Glow Contributors. See CONTRIBUTORS file.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <map>

extern std::map<float, float> GlobalMap;
std::map<float, float> GlobalMap;

extern "C" {
/// Simply write a global map to validate that static C++ constructors are
/// correctly called.
void libjit_WriteGlobalMap(float key, float value) { GlobalMap[key] = value; }
/// The following variables needs to be defined in libjit, because codegens look
/// for it.
size_t libjit_sizeTVar;
}
```
and a matching CMake change:
```cmake
if("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
    set(CLANG_BIN ${CMAKE_CXX_COMPILER})
  else()
    find_program(CLANG_BIN clang++)
  endif()
  if(NOT CLANG_BIN)
    message(SEND_ERROR "unable to find clang, cannot build the CPU runtime")
  endif()

  add_custom_command(
    OUTPUT ${GLOW_BINARY_DIR}/tests/mini_libjit.bc
    COMMAND ${CLANG_BIN} -emit-llvm -c -o ${GLOW_BINARY_DIR}/tests/mini_libjit.bc mini_libjit.cpp
    DEPENDS mini_libjit.cpp
    WORKING_DIRECTORY "${CMAKE_CURRENT_LIST_DIR}")

  file(MAKE_DIRECTORY ${GLOW_BINARY_DIR}/glow/CPU)
  add_custom_command(
    OUTPUT ${GLOW_BINARY_DIR}/glow/CPU/mini_libjit_bc.inc
    COMMAND include-bin "${GLOW_BINARY_DIR}/tests/mini_libjit.bc" "${GLOW_BINARY_DIR}/glow/CPU/mini_libjit_bc.inc"
    DEPENDS ${GLOW_BINARY_DIR}/tests/mini_libjit.bc
    WORKING_DIRECTORY "${CMAKE_CURRENT_LIST_DIR}")

  add_executable(CPUJITTest
                 CPUJITTest.cpp ${GLOW_BINARY_DIR}/glow/CPU/mini_libjit_bc.inc)
  target_link_libraries(CPUJITTest
                        PRIVATE
                          Backends
                          ExecutionEngine
                          Graph
                          Support
                          gtest
                          TestMain)
endif()
```

opti-mix commented:
@tlepley-cadence Ping! :-) And 🎄 🎅

opti-mix commented:
@tlepley-cadence Ping! :-)

tlepley-cadence commented:
I'm back !! :-) I was very busy with other things before Christmas, so sorry for not being very responsive ;-)

opti-mix commented:
@tlepley-cadence Do you still plan to land this?

tlepley-cadence commented:
@opti-mix yes, this is be my main task tomorrow to rework the test like you suggested. By the way, part of the PR repair the 'external backend' mechanics. I'll submit a PR for this particular subject separatly.

tlepley-cadence commented:
@opti-mix I finally reworked the PR and I have implemented the testing following your approach. I've cleaned a bit the testing code to make sure we can easily add other JIT tests if we want.

Regarding the the symbol linkage model you had detected on Mac (and that we had seen under Windows), I propose that you create another PR about it since I don't see this problem under Linux (I would then be complicated for me to make sure the PR is relevant). It's about the code snippet you have experimented in Pipeline.cpp (optimizeLLVMModule)

opti-mix commented:
@tlepley-cadence Thanks for updating the PR. Looks pretty good now!

> Regarding the other symbol linkage problem you had detected on Mac (and that we had seen under Windows), I propose that you create another PR because since I don't see this problem under Linux and it would then be complicated for me to make sure the PR is correct. It's about the code snippet you have experimented in Pipeline.cpp (optimizeLLVMModule)

Let's see what our CI will say about it. My concern is that if we merge this PR as is, it will break local builds for people building on MacOS or Windows due to this new test that you added.

tlepley-cadence commented:
> Let's see what our CI will say about it. My concern is that if we merge this PR as is, it will break local builds for people building on MacOS or Windows due to this new test that you added.

If you can create a branch derived from mine with an extra commit with the fix, and test it under MacOS, I can then cherry-pick your commit to incorporate it in this PR before we land it.

tlepley-cadence commented:
> Let's see what our CI will say about it.

It seems that the JIT test I've added that works well in my Linux environment fails in you CI

opti-mix commented:
@tlepley-cadence 

OK. CI reports some errors:

- Thread Sanitizer reports a lot of issues. Here are just some of them:

```
==================
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=13465)
    #0 operator new[](unsigned long) <null> (CPUJITTest+0x575cdd)
    #1 llvm::raw_ostream::SetBuffered() <null> (CPUJITTest+0x282a25e)
    #2 llvm::llvm_unreachable_internal(char const*, char const*, unsigned int) <null> (CPUJITTest+0x27da2a0)
    #3 llvm::orc::GlowJIT::addModule(std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >) /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/GlowJIT.cpp:232:3 (CPUJITTest+0x94f6c5)
    #4 glow::LLVMBackend::compileIRWithoutConstants(glow::IRFunction*) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:133:8 (CPUJITTest+0x9c44ce)
    #5 glow::LLVMBackend::compile(glow::Function*, glow::BackendOptions const&) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:156:20 (CPUJITTest+0x9c49c5)
    #6 glow::runtime::Provisioner::provision(std::vector<glow::runtime::DAG, std::allocator<glow::runtime::DAG> >&, glow::Module&, glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/Runtime/Provisioner/Provisioner.cpp:352:43 (CPUJITTest+0x8d8297)
    #7 glow::runtime::HostManager::addNetwork(std::unique_ptr<glow::Module, std::default_delete<glow::Module> >, glow::CompilationContext&, bool) /var/lib/jenkins/workspace/build/../lib/Runtime/HostManager/HostManager.cpp:221:28 (CPUJITTest+0x7ff5fc)
    #8 glow::ExecutionEngine::compile(glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:275:3 (CPUJITTest+0x59e878)
    #9 glow::ExecutionEngine::compile(glow::CompilationContext::CompilationMode) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:265:3 (CPUJITTest+0x59e6a9)
    #10 CPUJITTest_testCppConstructors_Test::TestBody() /var/lib/jenkins/workspace/build/../tests/unittests/CPUJITTest.cpp:141:6 (CPUJITTest+0x5780c3)
    #11 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:10 (CPUJITTest+0x7eba96)
    #12 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14 (CPUJITTest+0x7c328e)
    #13 testing::Test::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2508:5 (CPUJITTest+0x792ffb)
    #14 testing::TestInfo::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2684:11 (CPUJITTest+0x7942f9)
    #15 testing::TestSuite::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2816:28 (CPUJITTest+0x794dd0)
    #16 testing::internal::UnitTestImpl::RunAllTests() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:5338:44 (CPUJITTest+0x7a7af1)
    #17 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:10 (CPUJITTest+0x7f26a6)
    #18 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14 (CPUJITTest+0x7c783e)
    #19 testing::UnitTest::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:4925:10 (CPUJITTest+0x7a733b)
    #20 RUN_ALL_TESTS() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/include/gtest/gtest.h:2473:46 (CPUJITTest+0x7fcb4d)
    #21 main /var/lib/jenkins/workspace/build/../tests/unittests/TestMain.cpp:25:10 (CPUJITTest+0x7fcb03)

==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=13465)
    #0 operator delete[](void*) <null> (CPUJITTest+0x5762cd)
    #1 printSymbolizedStackTrace(llvm::StringRef, void**, int, llvm::raw_ostream&)::$_0::operator()() const <null> (CPUJITTest+0x2846c61)
    #2 llvm::llvm_unreachable_internal(char const*, char const*, unsigned int) <null> (CPUJITTest+0x27da2a0)
    #3 llvm::orc::GlowJIT::addModule(std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >) /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/GlowJIT.cpp:232:3 (CPUJITTest+0x94f6c5)
    #4 glow::LLVMBackend::compileIRWithoutConstants(glow::IRFunction*) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:133:8 (CPUJITTest+0x9c44ce)
    #5 glow::LLVMBackend::compile(glow::Function*, glow::BackendOptions const&) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:156:20 (CPUJITTest+0x9c49c5)
    #6 glow::runtime::Provisioner::provision(std::vector<glow::runtime::DAG, std::allocator<glow::runtime::DAG> >&, glow::Module&, glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/Runtime/Provisioner/Provisioner.cpp:352:43 (CPUJITTest+0x8d8297)
    #7 glow::runtime::HostManager::addNetwork(std::unique_ptr<glow::Module, std::default_delete<glow::Module> >, glow::CompilationContext&, bool) /var/lib/jenkins/workspace/build/../lib/Runtime/HostManager/HostManager.cpp:221:28 (CPUJITTest+0x7ff5fc)
    #8 glow::ExecutionEngine::compile(glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:275:3 (CPUJITTest+0x59e878)
    #9 glow::ExecutionEngine::compile(glow::CompilationContext::CompilationMode) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:265:3 (CPUJITTest+0x59e6a9)
    #10 CPUJITTest_testCppConstructors_Test::TestBody() /var/lib/jenkins/workspace/build/../tests/unittests/CPUJITTest.cpp:141:6 (CPUJITTest+0x5780c3)
    #11 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:10 (CPUJITTest+0x7eba96)
    #12 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14 (CPUJITTest+0x7c328e)
    #13 testing::Test::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2508:5 (CPUJITTest+0x792ffb)
    #14 testing::TestInfo::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2684:11 (CPUJITTest+0x7942f9)
    #15 testing::TestSuite::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2816:28 (CPUJITTest+0x794dd0)
    #16 testing::internal::UnitTestImpl::RunAllTests() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:5338:44 (CPUJITTest+0x7a7af1)
    #17 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:10 (CPUJITTest+0x7f26a6)
    #18 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14 (CPUJITTest+0x7c783e)
    #19 testing::UnitTest::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:4925:10 (CPUJITTest+0x7a733b)
    #20 RUN_ALL_TESTS() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/include/gtest/gtest.h:2473:46 (CPUJITTest+0x7fcb4d)
    #21 main /var/lib/jenkins/workspace/build/../tests/unittests/TestMain.cpp:25:10 (CPUJITTest+0x7fcb03)

    #1 printSymbolizedStackTrace(llvm::StringRef, void**, int, llvm::raw_ostream&)::$_0::operator()() const <null> (CPUJITTest+0x2846d52)
    #2 llvm::llvm_unreachable_internal(char const*, char const*, unsigned int) <null> (CPUJITTest+0x27da2a0)
    #3 llvm::orc::GlowJIT::addModule(std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >) /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/GlowJIT.cpp:232:3 (CPUJITTest+0x94f6c5)
    #4 glow::LLVMBackend::compileIRWithoutConstants(glow::IRFunction*) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:133:8 (CPUJITTest+0x9c44ce)
    #5 glow::LLVMBackend::compile(glow::Function*, glow::BackendOptions const&) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:156:20 (CPUJITTest+0x9c49c5)
    #6 glow::runtime::Provisioner::provision(std::vector<glow::runtime::DAG, std::allocator<glow::runtime::DAG> >&, glow::Module&, glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/Runtime/Provisioner/Provisioner.cpp:352:43 (CPUJITTest+0x8d8297)
    #7 glow::runtime::HostManager::addNetwork(std::unique_ptr<glow::Module, std::default_delete<glow::Module> >, glow::CompilationContext&, bool) /var/lib/jenkins/workspace/build/../lib/Runtime/HostManager/HostManager.cpp:221:28 (CPUJITTest+0x7ff5fc)
    #8 glow::ExecutionEngine::compile(glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:275:3 (CPUJITTest+0x59e878)
    #9 glow::ExecutionEngine::compile(glow::CompilationContext::CompilationMode) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:265:3 (CPUJITTest+0x59e6a9)
    #10 CPUJITTest_testCppConstructors_Test::TestBody() /var/lib/jenkins/workspace/build/../tests/unittests/CPUJITTest.cpp:141:6 (CPUJITTest+0x5780c3)
    #11 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:10 (CPUJITTest+0x7eba96)
    #12 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14 (CPUJITTest+0x7c328e)
    #13 testing::Test::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2508:5 (CPUJITTest+0x792ffb)
    #14 testing::TestInfo::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2684:11 (CPUJITTest+0x7942f9)
    #15 testing::TestSuite::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2816:28 (CPUJITTest+0x794dd0)
    #16 testing::internal::UnitTestImpl::RunAllTests() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:5338:44 (CPUJITTest+0x7a7af1)
    #17 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:10 (CPUJITTest+0x7f26a6)
    #18 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14 (CPUJITTest+0x7c783e)
    #19 testing::UnitTest::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:4925:10 (CPUJITTest+0x7a733b)
    #20 RUN_ALL_TESTS() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/include/gtest/gtest.h:2473:46 (CPUJITTest+0x7fcb4d)
    #21 main /var/lib/jenkins/workspace/build/../tests/unittests/TestMain.cpp:25:10 (CPUJITTest+0x7fcb03)


SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal (/var/lib/jenkins/workspace/build/tests/CPUJITTest+0x57622d) in operator delete(void*)
==================
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=13465)
    #0 operator delete(void*) <null> (CPUJITTest+0x57622d)
    #1 printSymbolizedStackTrace(llvm::StringRef, void**, int, llvm::raw_ostream&)::$_0::operator()() const <null> (CPUJITTest+0x2846d64)
    #2 llvm::llvm_unreachable_internal(char const*, char const*, unsigned int) <null> (CPUJITTest+0x27da2a0)
    #3 llvm::orc::GlowJIT::addModule(std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >) /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/GlowJIT.cpp:232:3 (CPUJITTest+0x94f6c5)
    #4 glow::LLVMBackend::compileIRWithoutConstants(glow::IRFunction*) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:133:8 (CPUJITTest+0x9c44ce)
    #5 glow::LLVMBackend::compile(glow::Function*, glow::BackendOptions const&) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:156:20 (CPUJITTest+0x9c49c5)
    #6 glow::runtime::Provisioner::provision(std::vector<glow::runtime::DAG, std::allocator<glow::runtime::DAG> >&, glow::Module&, glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/Runtime/Provisioner/Provisioner.cpp:352:43 (CPUJITTest+0x8d8297)
    #7 glow::runtime::HostManager::addNetwork(std::unique_ptr<glow::Module, std::default_delete<glow::Module> >, glow::CompilationContext&, bool) /var/lib/jenkins/workspace/build/../lib/Runtime/HostManager/HostManager.cpp:221:28 (CPUJITTest+0x7ff5fc)
    #8 glow::ExecutionEngine::compile(glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:275:3 (CPUJITTest+0x59e878)
    #9 glow::ExecutionEngine::compile(glow::CompilationContext::CompilationMode) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:265:3 (CPUJITTest+0x59e6a9)
    #10 CPUJITTest_testCppConstructors_Test::TestBody() /var/lib/jenkins/workspace/build/../tests/unittests/CPUJITTest.cpp:141:6 (CPUJITTest+0x5780c3)
    #11 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:10 (CPUJITTest+0x7eba96)
    #12 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14 (CPUJITTest+0x7c328e)
    #13 testing::Test::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2508:5 (CPUJITTest+0x792ffb)
    #14 testing::TestInfo::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2684:11 (CPUJITTest+0x7942f9)
    #15 testing::TestSuite::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2816:28 (CPUJITTest+0x794dd0)
    #16 testing::internal::UnitTestImpl::RunAllTests() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:5338:44 (CPUJITTest+0x7a7af1)
    #17 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:10 (CPUJITTest+0x7f26a6)
    #18 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14 (CPUJITTest+0x7c783e)
    #19 testing::UnitTest::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:4925:10 (CPUJITTest+0x7a733b)
    #20 RUN_ALL_TESTS() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/include/gtest/gtest.h:2473:46 (CPUJITTest+0x7fcb4d)
    #21 main /var/lib/jenkins/workspace/build/../tests/unittests/TestMain.cpp:25:10 (CPUJITTest+0x7fcb03)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal (/var/lib/jenkins/workspace/build/tests/CPUJITTest+0x57622d) in operator delete(void*)
==================
testing::TestSuite::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2816:28
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=13465)
    #0 operator new(unsigned long) <null> (CPUJITTest+0x575b4d)
    #1 llvm::formatv_object_base::parseFormatString(llvm::StringRef) <null> (CPUJITTest+0x27dce35)
    #2 llvm::llvm_unreachable_internal(char const*, char const*, unsigned int) <null> (CPUJITTest+0x27da2a0)
    #3 llvm::orc::GlowJIT::addModule(std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >) /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/GlowJIT.cpp:232:3 (CPUJITTest+0x94f6c5)
    #4 glow::LLVMBackend::compileIRWithoutConstants(glow::IRFunction*) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:133:8 (CPUJITTest+0x9c44ce)
    #5 glow::LLVMBackend::compile(glow::Function*, glow::BackendOptions const&) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:156:20 (CPUJITTest+0x9c49c5)
    #6 glow::runtime::Provisioner::provision(std::vector<glow::runtime::DAG, std::allocator<glow::runtime::DAG> >&, glow::Module&, glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/Runtime/Provisioner/Provisioner.cpp:352:43 (CPUJITTest+0x8d8297)
    #7 glow::runtime::HostManager::addNetwork(std::unique_ptr<glow::Module, std::default_delete<glow::Module> >, glow::CompilationContext&, bool) /var/lib/jenkins/workspace/build/../lib/Runtime/HostManager/HostManager.cpp:221:28 (CPUJITTest+0x7ff5fc)
    #8 glow::ExecutionEngine::compile(glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:275:3 (CPUJITTest+0x59e878)
    #9 glow::ExecutionEngine::compile(glow::CompilationContext::CompilationMode) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:265:3 (CPUJITTest+0x59e6a9)
    #10 CPUJITTest_testCppConstructors_Test::TestBody() /var/lib/jenkins/workspace/build/../tests/unittests/CPUJITTest.cpp:141:6 (CPUJITTest+0x5780c3)
    #11 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:10 (CPUJITTest+0x7eba96)
    #12 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14 (CPUJITTest+0x7c328e)
    #13 testing::Test::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2508:5 (CPUJITTest+0x792ffb)
    #14 testing::TestInfo::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2684:11 (CPUJITTest+0x7942f9)
    #15 testing::TestSuite::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2816:28 (CPUJITTest+0x794dd0)
    #16 testing::internal::UnitTestImpl::RunAllTests() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:5338:44 (CPUJITTest+0x7a7af1)
    #17 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:10 (CPUJITTest+0x7f26a6)
    #18 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14 (CPUJITTest+0x7c783e)
    #19 testing::UnitTest::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:4925:10 (CPUJITTest+0x7a733b)
    #20 RUN_ALL_TESTS() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/include/gtest/gtest.h:2473:46 (CPUJITTest+0x7fcb4d)
    #21 main /var/lib/jenkins/workspace/build/../tests/unittests/TestMain.cpp:25:10 (CPUJITTest+0x7fcb03)

```


- OpenCL build fails with this:

```
 #2 0x00000000025582c8 SignalHandler(int) (/var/lib/jenkins/workspace/build/tests/CPUJITTest+0x25582c8)
 #3 0x00007f3cc1c4c390 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x11390)
 #4 0x00007f3cc1005428 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x35428)
 #5 0x00007f3cc100702a abort (/lib/x86_64-linux-gnu/libc.so.6+0x3702a)
 #6 0x00000000024ea901 (/var/lib/jenkins/workspace/build/tests/CPUJITTest+0x24ea901)
 #7 0x0000000000728a79 llvm::cantFail(llvm::Error, char const*) /usr/lib/llvm-8/include/llvm/Support/Error.h:710:1
 #8 0x000000000074186c llvm::orc::GlowJIT::addModule(std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >) /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/GlowJIT.cpp:232:3
 #9 0x0000000000789c82 glow::LLVMBackend::compileIRWithoutConstants(glow::IRFunction*) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:133:3
#10 0x000000000078a095 glow::LLVMBackend::compile(glow::Function*, glow::BackendOptions const&) const /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/LLVMBackend.cpp:156:18
#11 0x00000000006f91e2 glow::runtime::Provisioner::provision(std::vector<glow::runtime::DAG, std::allocator<glow::runtime::DAG> >&, glow::Module&, glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/Runtime/Provisioner/Provisioner.cpp:352:43
#12 0x0000000000676a74 glow::runtime::HostManager::addNetwork(std::unique_ptr<glow::Module, std::default_delete<glow::Module> >, glow::CompilationContext&, bool) /var/lib/jenkins/workspace/build/../lib/Runtime/HostManager/HostManager.cpp:222:7
#13 0x00000000004de2e8 glow::ExecutionEngine::compile(glow::CompilationContext&) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:275:3
#14 0x00000000004de162 glow::ExecutionEngine::compile(glow::CompilationContext::CompilationMode) /var/lib/jenkins/workspace/build/../lib/ExecutionEngine/ExecutionEngine.cpp:266:1
#15 0x00000000004c6a24 CPUJITTest_testCppConstructors_Test::TestBody() /var/lib/jenkins/workspace/build/../tests/unittests/CPUJITTest.cpp:142:6
#16 0x000000000066c4be void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:3
#17 0x00000000006567cb void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14
#18 0x00000000006367e3 testing::Test::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2515:3
#19 0x0000000000637558 testing::TestInfo::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2687:12
#20 0x0000000000637c77 testing::TestSuite::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2815:44
#21 0x0000000000644d1a testing::internal::UnitTestImpl::RunAllTests() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:5337:24
#22 0x000000000066f18e bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2433:3
#23 0x0000000000658e2b bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:2469:14
#24 0x0000000000644846 testing::UnitTest::Run() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/src/gtest.cc:4925:10
#25 0x0000000000674701 RUN_ALL_TESTS() /var/lib/jenkins/workspace/build/../tests/googletest/googletest/include/gtest/gtest.h:2473:3
#26 0x00000000006746e4 main /var/lib/jenkins/workspace/build/../tests/unittests/TestMain.cpp:25:3
#27 0x00007f3cc0ff0830 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x20830)
#28 0x00000000004c6459 _start (/var/lib/jenkins/workspace/build/tests/CPUJITTest+0x4c6459)
```
 
All those tests are performed by our CI on a Linux machine.

opti-mix commented:
@tlepley-cadence 

> If you can create a branch derived from mine with an extra commit with the fix, and test it under MacOS, I can then cherry-pick your commit to incorporate it in this PR before we land it.

Yes, I could do it once you fix the current CI issues.

opti-mix commented on lib/LLVMIRCodeGen/GlowJIT.cpp:
```diff
@@ -160,24 +185,67 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
+#endif
+
+  // Run the static constructors and register static destructors.
+  cantFail(ctorRunner.runViaLayer(compileLayer_));
```
CI reports this line as the reason for a failure in the DEBUG build.
```
 #8 0x000000000073ddec llvm::orc::GlowJIT::addModule(std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >) /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/GlowJIT.cpp:232:3
```

tlepley-cadence replied:
Yes, that's what I was looking at. I had added `cantFail` when calling `ctorRunner.runViaLayer` to please my IDE, but looking at of the the examples using ORCv1, I see that they don't use `cantFail`: https://github.com/microsoft/DirectXShaderCompiler/blob/master/tools/lli/OrcLazyJIT.h#L118

Then, I'll simply remove it to see what is going on.
I'll also integrate you other patch to this PR directly.

tlepley-cadence commented:
@opti-mix It looks better in the CI now. Could you please test this branch on your side ?

opti-mix commented:
@tlepley-cadence 

> Could you please test this branch on your side ?

Just checked on MacOS and it works fine with LLVM7.

But it fails on the FB internal CI. It cannot find `LegacyLocalCXXRuntimeOverrides`. I guess we'll need to sprinkle some `#ifdefs`.

opti-mix commented:
@tlepley-cadence Could you try to use in `#ifdefs` the condition like 
```
#if LLVM_VERSION_MAJOR == 7 || FACEBOOK_INTERNAL
```

Let's see if this helps.

vuzelac-cadence commented:
About the testing - might be useful to add such static variable deconstructor testing.
@tlepley I remember that  in the middle of your development, we had a problem w/ program exit calling deconstructors incorrectly.

@opti-mix , somewhat tangential; would it be useful to have some real feature done using a static variable ? With minibatch-threads, I don't see static variables are any useful as they are not really shared across threads. However, might be interesting to improve e.g. instrument debug to dump to a file. Could protect that file using a global lock.

opti-mix commented:
@tlepley-cadence 

I can confirm that using 
```
#if LLVM_VERSION_MAJOR == 7 || FACEBOOK_INTERNAL
```
solves the compilation issues on FB's internal CI systems.

opti-mix commented:
@vuzelac-cadence 

> somewhat tangential; would it be useful to have some real feature done using a static variable ? With minibatch-threads, I don't see static variables are any useful as they are not really shared across threads. However, might be interesting to improve e.g. instrument debug to dump to a file. Could protect that file using a global lock. Anyway, we dump to files through the Host (Glow) but still...

There is not real need for using any static variables so far on our side. But since we'll have a small CPU JIT test once we land this PR, we could add more tests for using static variables as well. Having a test to make sure static variables can be used is a nice idea.

tlepley-cadence commented:
@vuzelac-cadence I don't remember this problem. But now that we have a minimal infrastructure for testing the JIT, we'll be able to create more tests are add additional tests once this PR has landed. I think this JIT story needs more testing and investigation independently from this PR (we for example still have this strange fprintf behavior under Windows (#3937). Also, Glow uses the old API of ORC. We'll certainly benefit migrating to the laster API at some point.

opti-mix commented on include/glow/LLVMIRCodeGen/GlowJIT.h:
```diff
@@ -52,18 +53,36 @@ class GlowJIT {
 #else
   std::shared_ptr<SymbolStringPool> SSP_;
   ExecutionSession ES_;
+  /// Handles symbols that are overriden by the JIT engine (needed to manage
```
overriden -> overridden

tlepley-cadence replied:
Fixed.

opti-mix commented on lib/LLVMIRCodeGen/GlowJIT.cpp:
```diff
@@ -160,24 +185,67 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
+  return MangledNameStream.str();
+}
+
+llvm::JITSymbol GlowJIT::findSymbol(const std::string name) {
```
`const std::string name` -> `const std::string &name`?

tlepley-cadence replied:
Fixed.
